exploration/0475-overriding-extending-namespacing.md

```diff
@@ -0,0 +1,334 @@
+  this might take the form of the existing resolution error
+- Addition of namespace to the data model for all relative items
+
+It is possible that the registry description will need to include slots for URL
+and default namespace name.
+
+### Potential Negatives
+
+This design is based on the assertion that implementors will provide an
+extension mechanism and that users will want to use that mechanism to install formatting
+or selection functionality.
+Any non-standard functions, options, option values, or expressions have the potential
+to be disruptive or fragmenting to the overall tooling or localization space.
+Any extension that is widely adopted would thus be better off in the default registry
+if at all possible.
+On the other hand, language- or platform-specific extensions can make MFv2 feel
+more "fluent" or consistent for users in a given environment.
+
+## Alternatives Considered
+
+_What other solutions are available?_
+_How do they compare against the requirements?_
+_What other properties they have?_
+
+### Choice of namespace separator
+
+The choice of the `:` character as a separator is, as noted above, a considered part
+of the design, but it is not required.
+Other characters are possible for the same purpose.
+
+A concern is that `name-char` reserves `:`, `-`, and `.`, which are probably the
+most likely candidates.
+Choosing one of these characters would require altering the `name-char` (and
+thus `name`) production.
+Moving to use `NCName` instead of `Nmtoken` as a basis for `name` would address
+the use of `:`.
+Moving to `-` or `.` would not require alterations of the syntax if we decided
+that namespacing is not formally a part of the specification but, rather, is
+just a convention.
+
+Of the remaining characters not currently in use, the most obvious candidates would
+be `/` and `~`.
+
+> Here are examples of the options:
+>
+> ```
+> {$today :ns:function  ns:option=foo}{+ns:a}{-ns:a}
+> {$today :ns-function  ns-option=foo}{+ns-a}{-ns-a}
+> {$today :ns.function  ns.option=foo}{+ns.a}{-ns.a}
+> {$today :ns/function  ns/option=foo}{+ns/a}{-ns/a}
+> {$today :ns~function  ns~option=foo}{+ns~a}{-ns~a}
+> ```
+
+### No namespacing
+
+Each implementation can install whatever additional functionality.
+It is up to the implementation to describe what is permitted and to check messages.
+Users will have to RTFM.
+
+- **+** Flexible
+- **-** Does not promote a healthy ecosystem of add-on packages
+- **-** Does not supply a mechanism for tooling to leverage
+
+### Reverse-domain-name namespacing
+
+Use `com.foo.bar.baz.Function` type naming for functions, options, or expressions.
+
+> ```
+> Today is {$today :com.example.foo.datetime dateStyle=short}
+> Today is {$today :datetime com.example.foo.skeleton=EEEEMd}
+> ```
+
+- **+** Familiarity. This is a familiar structure for developers.
+- **-** Verbose. The resulting names are long and difficult to parse visually
+
+### Namespacing sigils are customary not normative
```

Comment:
This would be my preferred solution. Are there significant negatives to it compared to the proposed solution?

Reply:
Why do you prefer this one?

IIUC, it doesn't address the requirement about machine-readability: _Translators and tools would like a machine-readable way to find out the names and option values for add-on packages_.

Additionally, messages can end up using different namespacing conventions, in particular in big projects. Coupled with the first issue, unifying them automatically/programmatically can be problematic.

Reply:
> Why do you prefer this one?

Because I find it simpler and lighter, and achieving the same results without explicitly adding a new sigil into the syntax.

> IIUC, it doesn't address the requirement about machine-readability: _Translators and tools would like a machine-readable way to find out the names and option values for add-on packages_.

As I understand it, any such machine-readability will need to rely on something like a function registry. My argument with respect to that is that equivalent results may be achieved by declaring in such a registry that a set of functionality uses the `moz-` prefix, rather than the `moz` namespace.

> Additionally, messages can end up using different namespacing conventions, in particular in big projects. Coupled with the first issue, unifying them automatically/programmatically can be problematic.

To reduce such, we could well include a SHOULD statement with a recommended sigil; this would almost certainly be followed by the vast majority of users.

Reply:
> As I understand it, any such machine-readability will need to rely on something like a function registry.

I don't think the registry should be relevant here. Instead, I'm imagining the AST for an annotation to look something like `{type: "function", ns: "moz", name: "plural", operand: ..., options: ...}`, which makes the namespace (`ns`) a first-class concept in the data model.

> To reduce such, we could well include a SHOULD statement with a recommended sigil; this would almost certainly be followed by the vast majority of users.

We may be putting a different amount of trust in the SHOULDs :)

Reply:
What @stasm said. There is no requirement that implementations be able to read the default registry file or implement the concept of a registry. They aren't required to provide an extension mechanism either. What this change does is provide a convenient syntactic route by which an implementation can have an extension mechanism and users can write messages that use it. This includes facilitating a translation toolchain downstream of the message.

We don't have to make the namespace/namespace prefix part of the syntax. If it were "just a convention", though, that would prevent us from ever formalizing it as a mechanism (since users might have created `moz-foo` functions that are not actually namespaced). Another option would be to reserve the prefix mechanism for future standardization.

Reply:
Ok, but what do you then do with a `ns: "moz"` without something like a registry?

Reply:
> Ok, but what do you then do with a `ns: "moz"` without something like a registry?

I'm looking at it from the PoV of the requirement: _Translators and tools would like a machine-readable way to find out the names and option values for add-on packages._ Having `ns: "moz"` in the data model means that it can be reliably identified by the tooling, e.g. for highlighting, "find all references", mass rename, etc.

Reply:
Reconsidered my stance here. Happy now for us to make the separator explicitly `:` in the syntax.